### DISCUSSION TITLE
Help me setup sslocal as a VPN

### DISCUSSION THREAD

**Discussion by enochc** [shadowsocks/shadowsocks-rust#1731] [General]
Forgive me, I'm new to shadowsocks and have never dug this deep into VPN's before. I'm trying to get this working as a VPN. I have a couple of challenges. First, how to configure sslocal to work? Here is my config I pass to the ssserver:
```json
{
    "server": "127.0.0.1",
    "outbound-bind-interface": "en0",
    "ipv6_first": false,
    "ipv6_only": false,
    "server_port": 8083,
    "mode": "tcp_and_udp",
    "method": "aes-256-gcm",
    "password": "XpqH2eVhDbG3aXCJ9h34U2r8H40Spen5xcRUzGv55t8=",
    "timeout": 300,
    "udp_timeout": 300,
    "udp_max_associations": 512,
    "nameserver": "1.1.1.1",
    "users": {
		"guest1": "21+Vu15EiavstNoKrbrxNXGVJbts0croDB49diF0tgc=",
		"guest2": "vLGQFajroBzh6QuO/P7lW1TN0NPgH07lyKs79io5myk="
    }
}
```
I start that process and it seems to startup just fine.
Then I start up sslocal with this config:
```json
{
    "server": "127.0.0.1",
    "server_port": 8083,
    "ipv6_first": false,
    "ipv6_only": false,
    "mode": "tcp_and_udp",
    "password": "XpqH2eVhDbG3aXCJ9h34U2r8H40Spen5xcRUzGv55t8=",
    "method": "aes-256-gcm",
    "timeout": 300,
    "udp_timeout": 300,
    "udp_max_associations": 512,
    "nameserver": "1.1.1.1",
    "local_address": "127.0.0.1",
    "local_port": 1081,
    "protocol": "tun"
}
```
I'm running on a Mac, this creates a tun interface called utun6, so I create a default route to send traffic to that new interface:
sudo route add default -interface utun6
Once that is run I can then see tons of crazy logging on both ssserver and sslocal. If I run 'ping example.com' I get a ping response that's resolved to an IP address. But if I run 'curl example.com' I get: curl: (52) Empty reply from server

I have connected an outline client to the sserver instance and it works. I have also connected a shadowsocksX-NG client to ssserver and that also works. What am I doing wrong with my sslocal service?

My second challenge, is how can I run this with aead-cipher-2022 enabled? I can run ssserver with "method": "2022-blake3-aes-256-gcm" in the config and it runs, but I can't get any client to actually connect and use it.

Any help would be greatly appreciated.

PS I did try: "curl -L -m 3 example.com"
and got this:  curl: (28) Operation timed out after 3006 milliseconds with 0 bytes received

**Comment by zonyitoo** [comment]
> Once that is run I can then see tons of crazy logging on both ssserver and sslocal.

The network packet sent from `ssserver` must have been routed back to `sslocal`.

The configuration key in your configuration for `ssserver` was wrong, it should be:

```json
{
    "outbound_bind_interface": "en0",
}
```

You could give another try, but if I remember it clearly, bind interface doesn't work like Linux on macOS.

**Reply by enochc** [reply]
I don't understand, that's exactly what I have in my sserver config. I have noticed that mac's tend to have 'en0' as the default outbound interface, whereas most Linux machines use eth0

**Reply by AaronChen0** [reply]
It should be `outbound_bind_interface`, not `outbound-bind-interface`.

**Reply by enochc** [reply]
Oh, I got it!! thanks, sorry for being slow.
